limitador-server/src/envoy_rls/server.rs

```diff
@@ -110,6 +112,37 @@ impl RateLimitService for MyRateLimiter {
     }
 }
 
+// create response headers per https://datatracker.ietf.org/doc/id/draft-polli-ratelimit-headers-03.html
+pub fn to_response_header(
+    counters: Vec<Counter>,
+    limited: Option<String>,
+) -> Vec<HeaderValue> {
+    let mut headers = Vec::new();
+
+    if let Some(name) = limited {
+        headers.push(HeaderValue { key: "X-RateLimit-Limit-Hit".to_string(), value: name });
+    }
+
+    counters.into_iter().for_each(|counter| {
+        let mut buf = String::with_capacity(40);
+        let limit = counter.limit().max_value();
+        let window = counter.limit().seconds();
+        buf.push_str(format!("{}, {};w={}", limit, limit, window).as_str());
+
+        if let Some(name) = counter.limit().name() {
+            buf.push_str(format!(";name=\"{}\"", name).as_str());
+        }
+        headers.push(HeaderValue { key: "X-RateLimit-Limit".to_string(), value: buf });
+        if let Some(remaining) = counter.remaining() {
+            headers.push(HeaderValue { key: "X-RateLimit-Remaining".to_string(), value: format!("{}", remaining) });
+        }
+        if let Some(duration) = counter.expires_in() {
```

Comment:
If you look at how the check is performed, i.e. [for redis](https://github.com/Kuadrant/limitador/blob/main/limitador/src/storage/redis/redis_async.rs#L59), you'll see that the actual value is checked against, then dropped on the floor. 

Probably adding a "seam" that would take a Vec of `&mut Counter` so to update the value in the instance alongside doing the check _and_, again in the case of Redis, query the `TTL` for all these counters and updating the `expires_in` field of each `Counter` would be a first stab at this. This probably explains my other comment about creating an additional method as, depending on the counter storage this will have non neglectable overhead.

We probably could add another script that compounds these operations instead of doing the `MGET` as it is currently done and do the update while doing a `MGET/MTTL` of all the values in one go. I don't know what infinispan provides as options to address this, but a O(2n) would be initially acceptable, as we don't "ship" that anyways. 

To TL;DR I'd go for a naive approach initially and have the support be configurable, so that the overhead isn't imposed on everyone initially.